File: lingji-gateway-node/src/handler/files.js
import crypto from 'node:crypto';
import fs from 'node:fs';
import path from 'node:path';
import { randomUUID } from 'node:crypto';
import Busboy from 'busboy';

function randomToken(bytes = 16) {
  return crypto.randomBytes(bytes).toString('hex');
}

export function sanitizeUploadFilename(name) {
  const base = path.basename(String(name || '').trim());
  if (!base || base === '.' || base === '..') return 'upload';
  return base;
}

export class FileStore {
  /** @param {import('../config.js').GatewayConfig} cfg */
  constructor(cfg) {
    this.cfg = cfg;
    /** @type {Map<string, object>} */
    this.files = new Map();
    this._stop = false;
    this._timer = setInterval(() => this.purgeExpired(), 5 * 60 * 1000);
    fs.mkdirSync(cfg.fileStoreDir, { recursive: true, mode: 0o700 });
  }

  stop() {
    this._stop = true;
    clearInterval(this._timer);
  }

  purgeExpired() {
    const now = Date.now();
    for (const [id, f] of this.files) {
      if (now > f.expiresAt) {
        fs.unlink(f.path, () => {});
        this.files.delete(id);
      }
    }
  }
}

export class FilesHandler {
  /** @param {import('../config.js').GatewayConfig} cfg */
  constructor(cfg) {
    this.cfg = cfg;
    this.store = new FileStore(cfg);
  }

  authOK(req) {
    if (!this.cfg.authToken) return true;
    const auth = req.headers.authorization;
    if (auth === `Bearer ${this.cfg.authToken}`) return true;
    const url = new URL(req.url, 'http://localhost');
    return url.searchParams.get('token') === this.cfg.authToken;
  }

  downloadAuthOK(req, f) {
    const url = new URL(req.url, 'http://localhost');
    const tok = url.searchParams.get('token');
    if (tok) {
      return tok === f.downloadTok || tok === this.cfg.authToken;
    }
    return this.authOK(req);
  }

  maxBytes() {
    return this.cfg.fileMaxSizeBytes > 0
      ? this.cfg.fileMaxSizeBytes
      : 50 * 1024 * 1024;
  }

  /** @param {import('node:http').IncomingMessage} req @param {import('node:http').ServerResponse} res */
  async handle(req, res) {
    const url = new URL(req.url, 'http://localhost');
    const sub = url.pathname.replace(/^\/files\/?/, '');

    if (!sub) {
      if (req.method !== 'POST') {
        res.writeHead(405);
        res.end('method not allowed');
        return;
      }
      return this.handleUpload(req, res);
    }

    if (req.method !== 'GET') {
      res.writeHead(405);
      res.end('method not allowed');
      return;
    }
    return this.handleDownload(req, res, sub.split('/')[0]);
  }

  /** @param {import('node:http').IncomingMessage} req @param {import('node:http').ServerResponse} res */
  handleUpload(req, res) {
    if (!this.authOK(req)) {
      res.writeHead(401);
      res.end('unauthorized');
      return;
    }

    const maxBytes = this.maxBytes();
    const busboy = Busboy({
      headers: req.headers,
      limits: { fileSize: maxBytes + 1024 },
    });

    let name = 'download.bin';
    let altName = '';
    /** @type {import('node:stream').Writable | null} */
    let dest = null;
    let destPath = '';
    let written = 0;
    let fileId = '';
    let dlTok = '';
    let mimeType = 'application/octet-stream';
    let failed = false;

    busboy.on('field', (fieldname, val) => {
      if (fieldname === 'name') altName = val;
    });

    busboy.on('file', (fieldname, file, info) => {
      if (fieldname !== 'file' || failed) {
        file.resume();
        return;
      }
      name = sanitizeUploadFilename(info.filename || '');
      if (name === 'upload' && !info.filename) name = 'download.bin';
      if (altName) name = sanitizeUploadFilename(altName);
      mimeType = info.mimeType || 'application/octet-stream';

      fileId = randomUUID();
      dlTok = randomToken(16);
      destPath = path.join(this.cfg.fileStoreDir, fileId);
      dest = fs.createWriteStream(destPath, { mode: 0o600 });

      file.on('data', (chunk) => {
        written += chunk.length;
        if (written > maxBytes) {
          failed = true;
          file.resume();
        }
      });

      file.pipe(dest);
    });

    busboy.on('error', () => {
      failed = true;
      if (destPath) fs.unlink(destPath, () => {});
      if (!res.headersSent) {
        res.writeHead(413);
        res.end('payload too large or invalid multipart');
      }
    });

    busboy.on('finish', () => {
      if (failed || !fileId || !dest) {
        if (destPath) fs.unlink(destPath, () => {});
        if (!res.headersSent) {
          res.writeHead(400);
          res.end('missing file field');
        }
        return;
      }

      const entry = {
        id: fileId,
        name,
        mime: mimeType,
        sizeBytes: written,
        path: destPath,
        downloadTok: dlTok,
        expiresAt: Date.now() + this.cfg.fileTtlMs,
        maxDownloads: this.cfg.fileMaxDownloads || 10,
        downloads: 0,
      };
      this.store.files.set(fileId, entry);

      const body = {
        file_id: fileId,
        name,
        size_bytes: written,
        mime: mimeType,
        download_path: `/files/${fileId}?token=${dlTok}`,
      };
      res.writeHead(200, { 'Content-Type': 'application/json' });
      res.end(JSON.stringify(body));
    });

    req.pipe(busboy);
  }

  /** @param {import('node:http').IncomingMessage} req @param {import('node:http').ServerResponse} res @param {string} fileId */
  handleDownload(req, res, fileId) {
    const f = this.store.files.get(fileId);
    if (!f || Date.now() > f.expiresAt) {
      res.writeHead(404);
      res.end('not found');
      return;
    }
    if (!this.downloadAuthOK(req, f)) {
      res.writeHead(401);
      res.end('unauthorized');
      return;
    }
    if (f.downloads >= f.maxDownloads) {
      res.writeHead(410);
      res.end('download limit exceeded');
      return;
    }
    f.downloads += 1;
    res.writeHead(200, {
      'Content-Type': f.mime,
      'Content-Disposition': `attachment; filename="${f.name}"`,
    });
    fs.createReadStream(f.path).pipe(res);
  }
}

File: lingji-gateway/handler/files.go
package handler

import (
	"crypto/rand"
	"encoding/hex"
	"encoding/json"
	"fmt"
	"io"
	"log"
	"mime"
	"net/http"
	"os"
	"path/filepath"
	"strings"
	"sync"
	"time"

	"github.com/AUrlius/lingji-gateway/config"
	"github.com/google/uuid"
)

const defaultMaxFileBytes = 50 * 1024 * 1024 // 50MB MVP

type storedFile struct {
	ID           string
	Name         string
	Mime         string
	SizeBytes    int64
	Path         string
	DownloadTok  string
	ExpiresAt    time.Time
	Downloads    int
	MaxDownloads int
}

// FileStore 临时文件存储（磁盘 + TTL 清理）
type FileStore struct {
	mu     sync.RWMutex
	files  map[string]*storedFile
	cfg    *config.Config
	stopCh chan struct{}
}

func NewFileStore(cfg *config.Config) *FileStore {
	dir := cfg.FileStoreDir
	if dir == "" {
		dir = filepath.Join(os.TempDir(), "lingji-files")
	}
	if err := os.MkdirAll(dir, 0o700); err != nil {
		log.Printf("[Files] 创建存储目录失败: %v", err)
	}

	s := &FileStore{
		files:  make(map[string]*storedFile),
		cfg:    cfg,
		stopCh: make(chan struct{}),
	}
	go s.cleanupLoop()
	return s
}

func (s *FileStore) Stop() {
	close(s.stopCh)
}

func (s *FileStore) cleanupLoop() {
	ticker := time.NewTicker(5 * time.Minute)
	defer ticker.Stop()
	for {
		select {
		case <-s.stopCh:
			return
		case <-ticker.C:
			s.purgeExpired()
		}
	}
}

func (s *FileStore) purgeExpired() {
	now := time.Now()
	s.mu.Lock()
	defer s.mu.Unlock()
	for id, f := range s.files {
		if now.After(f.ExpiresAt) {
			_ = os.Remove(f.Path)
			delete(s.files, id)
		}
	}
}

func randomToken(n int) (string, error) {
	b := make([]byte, n)
	if _, err := rand.Read(b); err != nil {
		return "", err
	}
	return hex.EncodeToString(b), nil
}

type uploadResponse struct {
	FileID       string `json:"file_id"`
	Name         string `json:"name"`
	SizeBytes    int64  `json:"size_bytes"`
	Mime         string `json:"mime"`
	DownloadPath string `json:"download_path"`
}

// FilesHandler HTTP 临时文件上传/下载
type FilesHandler struct {
	store  *FileStore
	config *config.Config
}

func NewFilesHandler(cfg *config.Config) *FilesHandler {
	return &FilesHandler{
		store:  NewFileStore(cfg),
		config: cfg,
	}
}

func (h *FilesHandler) authOK(r *http.Request) bool {
	if h.config.AuthToken == "" {
		return true
	}
	if auth := r.Header.Get("Authorization"); auth == "Bearer "+h.config.AuthToken {
		return true
	}
	if r.URL.Query().Get("token") == h.config.AuthToken {
		return true
	}
	return false
}

func (h *FilesHandler) downloadAuthOK(r *http.Request, f *storedFile) bool {
	if tok := r.URL.Query().Get("token"); tok != "" {
		return tok == f.DownloadTok || (h.config.AuthToken != "" && tok == h.config.AuthToken)
	}
	return h.authOK(r)
}

func (h *FilesHandler) maxBytes() int64 {
	if h.config.FileMaxSizeBytes > 0 {
		return h.config.FileMaxSizeBytes
	}
	return defaultMaxFileBytes
}

func (h *FilesHandler) ttl() time.Duration {
	if h.config.FileTTL > 0 {
		return h.config.FileTTL
	}
	return time.Hour
}

func (h *FilesHandler) maxDownloads() int {
	if h.config.FileMaxDownloads > 0 {
		return h.config.FileMaxDownloads
	}
	return 10
}

func (h *FilesHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
	path := strings.TrimPrefix(r.URL.Path, "/files")
	path = strings.TrimPrefix(path, "/")
	if path == "" {
		if r.Method != http.MethodPost {
			http.Error(w, "method not allowed", http.StatusMethodNotAllowed)
			return
		}
		h.handleUpload(w, r)
		return
	}
	if r.Method != http.MethodGet {
		http.Error(w, "method not allowed", http.StatusMethodNotAllowed)
		return
	}
	h.handleDownload(w, r, path)
}

func sanitizeUploadFilename(name string) string {
	name = filepath.Base(strings.TrimSpace(name))
	if name == "" || name == "." || name == ".." {
		return "upload"
	}
	return name
}

func (h *FilesHandler) handleUpload(w http.ResponseWriter, r *http.Request) {
	if !h.authOK(r) {
		http.Error(w, "unauthorized", http.StatusUnauthorized)
		return
	}

	maxBytes := h.maxBytes()
	r.Body = http.MaxBytesReader(w, r.Body, maxBytes+1024)

	if err := r.ParseMultipartForm(maxBytes + 1024); err != nil {
		http.Error(w, "payload too large or invalid multipart", http.StatusRequestEntityTooLarge)
		return
	}

	file, header, err := r.FormFile("file")
	if err != nil {
		http.Error(w, "missing file field", http.StatusBadRequest)
		return
	}
	defer file.Close()

	name := sanitizeUploadFilename(header.Filename)
	if name == "upload" && header.Filename == "" {
		name = "download.bin"
	}
	if alt := r.FormValue("name"); alt != "" {
		name = sanitizeUploadFilename(alt)
	}

	id := uuid.New().String()
	dlTok, err := randomToken(16)
	if err != nil {
		http.Error(w, "internal error", http.StatusInternalServerError)
		return
	}

	dir := h.store.cfg.FileStoreDir
	if dir == "" {
		dir = filepath.Join(os.TempDir(), "lingji-files")
	}
	_ = os.MkdirAll(dir, 0o700)
	destPath := filepath.Join(dir, id)

	dest, err := os.Create(destPath)
	if err != nil {
		http.Error(w, "storage error", http.StatusInternalServerError)
		return
	}
	written, err := io.Copy(dest, file)
	dest.Close()
	if err != nil {
		_ = os.Remove(destPath)
		http.Error(w, "upload failed", http.StatusInternalServerError)
		return
	}
	if written > maxBytes {
		_ = os.Remove(destPath)
		http.Error(w, fmt.Sprintf("file exceeds limit (%d bytes)", maxBytes), http.StatusRequestEntityTooLarge)
		return
	}

	mimeType := header.Header.Get("Content-Type")
	if mimeType == "" || mimeType == "application/octet-stream" {
		mimeType = mime.TypeByExtension(filepath.Ext(name))
	}
	if mimeType == "" {
		mimeType = "application/octet-stream"
	}

	entry := &storedFile{
		ID:           id,
		Name:         name,
		Mime:         mimeType,
		SizeBytes:    written,
		Path:         destPath,
		DownloadTok:  dlTok,
		ExpiresAt:    time.Now().Add(h.ttl()),
		MaxDownloads: h.maxDownloads(),
	}

	h.store.mu.Lock()
	h.store.files[id] = entry
	h.store.mu.Unlock()

	resp := uploadResponse{
		FileID:       id,
		Name:         name,
		SizeBytes:    written,
		Mime:         mimeType,
		DownloadPath: fmt.Sprintf("/files/%s?token=%s", id, dlTok),
	}
	w.Header().Set("Content-Type", "application/json")
	_ = json.NewEncoder(w).Encode(resp)
}

func (h *FilesHandler) handleDownload(w http.ResponseWriter, r *http.Request, fileID string) {
	h.store.mu.RLock()
	f, ok := h.store.files[fileID]
	h.store.mu.RUnlock()
	if !ok || time.Now().After(f.ExpiresAt) {
		http.Error(w, "not found", http.StatusNotFound)
		return
	}
	if !h.downloadAuthOK(r, f) {
		http.Error(w, "unauthorized", http.StatusUnauthorized)
		return
	}

	h.store.mu.Lock()
	if f.Downloads >= f.MaxDownloads {
		h.store.mu.Unlock()
		http.Error(w, "download limit exceeded", http.StatusGone)
		return
	}
	f.Downloads++
	h.store.mu.Unlock()

	w.Header().Set("Content-Type", f.Mime)
	w.Header().Set("Content-Disposition", fmt.Sprintf(`attachment; filename="%s"`, f.Name))
	http.ServeFile(w, r, f.Path)
}
